File: algorithms/java/src/test/java/leetcode/LC367ValidPerfectSquare.java
package leetcode;

import org.junit.jupiter.api.Test;


import static org.junit.jupiter.api.Assertions.assertEquals;

/**
 * https://leetcode.com/problems/valid-perfect-square/
 * #easy
 */
public final class LC367ValidPerfectSquare {
    public final class Solution {
        public boolean isPerfectSquare(final int num) {
            long a = 0, z = 1 + num / 2;
            while (a <= z) {
                long m = (a + z) / 2;
                long mm = m * m;
                if (mm == num) {
                    return true;
                }
                if (mm > num) {
                    z = m - 1;
                } else {
                    a = m + 1;
                }
            }
            return false;
        }
    }

    @Test
    public void test_1() throws Exception {
        assertEquals(true, new Solution().isPerfectSquare(1));
    }

    @Test
    public void test_14() throws Exception {
        assertEquals(false, new Solution().isPerfectSquare(14));
    }

    @Test
    public void test_16() throws Exception {
        assertEquals(true, new Solution().isPerfectSquare(16));
    }

    @Test
    public void test_maximum() throws Exception {
        assertEquals(false, new Solution().isPerfectSquare(Integer.MAX_VALUE));
    }
}

File: algorithms/code/leetcode/lc367_valid_perfect_square/lc367_valid_perfect_square.rs
// https://leetcode.com/problems/valid-perfect-square/
// #easy

pub struct Solution;

impl Solution {
    pub fn is_perfect_square(num: i32) -> bool {
        let mut a = 0;
        let mut z = 1 + num / 2;
        while a <= z {
            let m = (a + z) / 2;
            let mm = m as i64 * m as i64;
            if mm == num as i64 {
                return true;
            }
            if mm > num as i64 {
                z = m - 1;
            } else {
                a = m + 1;
            }
        }
        false
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_1() {
        assert!(Solution::is_perfect_square(1));
    }

    #[test]
    fn test_14() {
        assert!(!Solution::is_perfect_square(14));
    }

    #[test]
    fn test_16() {
        assert!(Solution::is_perfect_square(16));
    }

    #[test]
    fn test_maximum() {
        assert!(!Solution::is_perfect_square(i32::MAX));
    }
}
